File: practice/taser_diy.py
import RPi.GPIO as GPIO
import time

# Set GPIO pin numbers
MOSFET1_PIN = 17
MOSFET2_PIN = 18
SWITCH_PIN = 27

# Set GPIO mode and pull-up for the switch pin
GPIO.setmode(GPIO.BCM)
GPIO.setup(SWITCH_PIN, GPIO.IN, pull_up_down=GPIO.PUD_UP)

# Set MOSFET pins as outputs
GPIO.setup(MOSFET1_PIN, GPIO.OUT)
GPIO.setup(MOSFET2_PIN, GPIO.OUT)

def switch_pressed(channel):
    # Turn MOSFET 1 on, MOSFET 2 off
    GPIO.output(MOSFET1_PIN, GPIO.HIGH)
    GPIO.output(MOSFET2_PIN, GPIO.LOW)

def switch_released(channel):
    # Turn MOSFET 1 off, MOSFET 2 on
    GPIO.output(MOSFET1_PIN, GPIO.LOW)
    GPIO.output(MOSFET2_PIN, GPIO.HIGH)

# Register callbacks for switch events
GPIO.add_event_detect(SWITCH_PIN, GPIO.FALLING, callback=switch_pressed, bouncetime=200)
GPIO.add_event_detect(SWITCH_PIN, GPIO.RISING, callback=switch_released, bouncetime=200)

try:
    while True:
        # Keep the program running
        time.sleep(1)

except KeyboardInterrupt:
    # Clean up GPIO
    GPIO.cleanup()

File: practice/taser_diy.cpp
#include <Arduino.h>

const int MOSFET1_PIN = 2;  // Connect MOSFET 1 to digital pin 2
const int MOSFET2_PIN = 3;  // Connect MOSFET 2 to digital pin 3
const int SWITCH_PIN = 4;   // Connect switch to digital pin 4

enum SwitchState {
  WAITING_FOR_PRESS,
  MOSFETS_ON,
  WAITING_FOR_RELEASE,
  MOSFETS_OFF
};

SwitchState switchState = WAITING_FOR_PRESS;

void setup() {
  pinMode(MOSFET1_PIN, OUTPUT);
  pinMode(MOSFET2_PIN, OUTPUT);
  pinMode(SWITCH_PIN, INPUT_PULLUP);  // Enable internal pull-up resistor for the switch
}

void loop() {
  switch (switchState) {
    case WAITING_FOR_PRESS:
      if (digitalRead(SWITCH_PIN) == LOW) {
        switchState = MOSFETS_ON;
        digitalWrite(MOSFET1_PIN, HIGH);
        digitalWrite(MOSFET2_PIN, LOW);
      }
      break;
      
    case MOSFETS_ON:
      if (digitalRead(SWITCH_PIN) == HIGH) {
        switchState = WAITING_FOR_RELEASE;
      }
      break;
      
    case WAITING_FOR_RELEASE:
      if (digitalRead(SWITCH_PIN) == HIGH) {
        switchState = MOSFETS_OFF;
        digitalWrite(MOSFET1_PIN, LOW);
        digitalWrite(MOSFET2_PIN, LOW);
      }
      break;
      
    case MOSFETS_OFF:
      if (digitalRead(SWITCH_PIN) == LOW) {
        switchState = WAITING_FOR_PRESS;
      }
      break;
  }
}
